File: bridge/python/src/destack/protocol/connection/connection.py
from __future__ import annotations

from collections.abc import Callable
from dataclasses import replace

from ..codec import decode_frame, encode_frame
from destack._generated.protocol.defaults import (
    client_descriptor,
    protocol_limits,
    protocol_range,
)
from destack._generated.protocol.envelope import (
    ProtocolMessageNotification,
    ProtocolMessageRequest,
    ProtocolMessageResponse,
    ProtocolNotification,
    ProtocolRequest,
    ProtocolResponse,
    RequestId,
    RequestOptions,
)
from destack._generated.protocol.handshake import (
    ClientDescriptor,
    HandshakeRequest,
    HandshakeResponse,
    ProtocolLimits,
)
from destack._generated.protocol.request import (
    WorkspaceRequest,
    WorkspaceRequestHandshake,
)
from destack._generated.protocol.response import (
    WorkspaceResponse,
    WorkspaceResponseError,
    WorkspaceResponseHandshake,
)
from destack._generated.protocol.version import ProtocolRange
from .error import ProtocolRequestError
from .payload.receiver import PayloadReceiver
from .transport import Transport, WebSocketTransport

NotificationHandler = Callable[[ProtocolNotification], None]


class Connection:
    """Synchronous connection to one workspace protocol endpoint."""

    def __init__(
        self,
        transport: Transport,
        on_notification: NotificationHandler | None = None,
    ) -> None:
        self._transport = transport
        self._on_notification = on_notification
        self._next_request_id = 1
        self._handshake: HandshakeResponse | None = None
        self._is_closed = False

    @classmethod
    def connect_websocket(
        cls,
        url: str,
        on_notification: NotificationHandler | None = None,
        client: ClientOptions | None = None,
    ) -> Connection:
        """Connect to one WebSocket workspace endpoint."""

        transport = WebSocketTransport.connect(url)
        connection = cls(transport=transport, on_notification=on_notification)
        connection.handshake(client)

        return connection

    @property
    def handshake_response(self) -> HandshakeResponse | None:
        """Return the negotiated handshake response, if any."""

        return self._handshake

    def handshake(self, options: ClientOptions | None = None) -> HandshakeResponse:
        """Negotiate the workspace protocol for this connection."""

        if self._handshake is not None:
            return self._handshake

        options = options or ClientOptions()
        request = WorkspaceRequestHandshake(
            handshake=HandshakeRequest(
                protocol=options.protocol or protocol_range,
                client=options.client_descriptor(),
                limits=options.limits or protocol_limits,
            ),
        )
        response = self.request(request)

        if isinstance(response, WorkspaceResponseHandshake):
            self._handshake = response.handshake
            return response.handshake

        if isinstance(response, WorkspaceResponseError):
            raise ProtocolRequestError.from_protocol(response.error)

        raise ProtocolRequestError(f"unexpected handshake response: {response.kind}")

    def request(
        self,
        payload: WorkspaceRequest,
        options: RequestOptions | None = None,
    ) -> WorkspaceResponse:
        """Send one workspace request and wait for its response."""

        if self._is_closed:
            raise ProtocolRequestError("workspace connection is closed")

        request_id = self._next_request_id
        self._next_request_id += 1
        request_options = options or RequestOptions(
            timeout_ms=None, priority=None, trace_id=None
        )
        request = ProtocolRequest(
            id=RequestId(field_0=request_id),
            options=request_options,
            payload=payload,
        )
        message = ProtocolMessageRequest(request=request)

        self._transport.send(encode_frame(message))

        return self._receive_response(request_id)

    def close(self) -> None:
        """Close this connection."""

        if self._is_closed:
            return

        self._is_closed = True
        self._transport.close()

    def _receive_response(self, request_id: int) -> WorkspaceResponse:
        """Receive messages until this request receives its response."""

        payloads = PayloadReceiver()
        response: WorkspaceResponse | None = None

        while response is None:
            message = decode_frame(self._transport.receive())

            if isinstance(message, ProtocolMessageResponse):
                response = self._response_for_id(message.response, request_id)
                continue

            if isinstance(message, ProtocolMessageNotification):
                is_payload = payloads.ingest_notification(message.notification)
                if not is_payload and self._on_notification is not None:
                    self._on_notification(message.notification)
                continue

            raise ProtocolRequestError("workspace connection received a request")

        response = payloads.resolve(response)
        if response is None:
            response = self._receive_payloads(request_id, payloads)

        if isinstance(response, WorkspaceResponseError):
            raise ProtocolRequestError.from_protocol(response.error)

        return response

    def _receive_payloads(
        self,
        request_id: int,
        payloads: PayloadReceiver,
    ) -> WorkspaceResponse:
        """Receive payload chunks required by one response."""

        while True:
            message = decode_frame(self._transport.receive())

            if isinstance(message, ProtocolMessageNotification):
                is_payload = payloads.ingest_notification(message.notification)
                if not is_payload and self._on_notification is not None:
                    self._on_notification(message.notification)
            elif isinstance(message, ProtocolMessageResponse):
                self._response_for_id(message.response, request_id)
                raise ProtocolRequestError(
                    "workspace connection received duplicate response"
                )
            else:
                raise ProtocolRequestError("workspace connection received a request")

            response = payloads.resolve_required()
            if response is not None:
                return response

    def _response_for_id(
        self,
        response: ProtocolResponse,
        request_id: int,
    ) -> WorkspaceResponse:
        """Return one response payload after checking its request id."""

        if response.id.field_0 != request_id:
            raise ProtocolRequestError(
                f"workspace connection received response id {response.id.field_0}, expected {request_id}"
            )

        return response.payload


class ClientOptions:
    """Options for workspace protocol negotiation."""

    def __init__(
        self,
        protocol: ProtocolRange | None = None,
        limits: ProtocolLimits | None = None,
        client: ClientDescriptor | None = None,
        name: str | None = None,
        version: str | None = None,
        build: str | None = None,
    ) -> None:
        self.protocol = protocol
        self.limits = limits
        self.client = client
        self.name = name
        self.version = version
        self.build = build

    def client_descriptor(self) -> ClientDescriptor:
        """Return the effective client descriptor."""

        descriptor = self.client or client_descriptor

        return replace(
            descriptor,
            name=self.name or descriptor.name,
            version=self.version or descriptor.version,
            build=self.build if self.build is not None else descriptor.build,
        )


def connect_endpoint(
    url: str,
    on_notification: NotificationHandler | None = None,
    client: ClientOptions | None = None,
) -> Connection:
    """Connect to one WebSocket workspace endpoint."""

    return Connection.connect_websocket(
        url=url,
        on_notification=on_notification,
        client=client,
    )

File: bridge/typescript/src/protocol/connection/connection.ts
import { decodeFrame, encodeFrame } from "../codec.js";
import {
    clientDescriptor,
    protocolLimits,
    protocolRange,
} from "../../_generated/protocol/defaults.js";
import type {
    ClientDescriptor,
    HandshakeResponse,
    ProtocolLimits,
} from "../../_generated/protocol/handshake.js";
import type {
    ProtocolMessage,
    ProtocolNotification,
    RequestOptions,
    ProtocolResponse,
} from "../../_generated/protocol/envelope.js";
import type { PayloadChunkNotification } from "../../_generated/protocol/payload.js";
import { WorkspaceRequest } from "../../_generated/protocol/request.js";
import type { WorkspaceResponse } from "../../_generated/protocol/response.js";
import type { ProtocolRange } from "../../_generated/protocol/version.js";
import { checkedNumber } from "./integer.js";
import { PayloadReceiver } from "./payload/receiver.js";
import type { Transport } from "./transport.js";
import { WebSocketTransport } from "./transport.js";

/** Handler for protocol notifications. */
export type NotificationHandler = (notification: ProtocolNotification) => void;

/** Options for workspace protocol negotiation. */
export type ClientOptions = {
    /** Supported protocol range on the client. */
    readonly protocol?: ProtocolRange;
    /** Requested protocol limits. */
    readonly limits?: ProtocolLimits;
    /** Client descriptor overrides. */
    readonly client?: Partial<ClientDescriptor>;
};

/** Options for one workspace protocol connection. */
export type ConnectionOptions = {
    /** Handle notifications sent by the workspace. */
    readonly onNotification?: NotificationHandler;
    /** Protocol negotiation options. */
    readonly client?: ClientOptions;
};

/** Pending response state for one active protocol request. */
type PendingResponse = {
    readonly resolve: (response: WorkspaceResponse) => void;
    readonly reject: (error: Error) => void;
    readonly payloads: PayloadReceiver;
    response?: WorkspaceResponse;
};

/** Async connection to one workspace protocol endpoint. */
export class Connection {
    readonly #transport: Transport;
    readonly #onNotification: NotificationHandler | undefined;
    readonly #pending = new Map<
        number,
        PendingResponse
    >();
    #nextRequestId = 1;
    #isClosed = false;
    #handshake: HandshakeResponse | undefined;
    #requests = Promise.resolve();

    /** Create one connection over an open binary transport. */
    constructor(transport: Transport, options: ConnectionOptions = {}) {
        this.#transport = transport;
        this.#onNotification = options.onNotification;
        void this.#receive();
    }

    /** Connect to one WebSocket workspace endpoint. */
    static async connectWebSocket(
        url: string | URL,
        options: ConnectionOptions = {},
    ): Promise<Connection> {
        const transport = await WebSocketTransport.connect(url);
        const connection = new Connection(transport, options);
        await connection.handshake(options.client);

        return connection;
    }

    /** Return the negotiated handshake response, if any. */
    get handshakeResponse(): HandshakeResponse | undefined {
        return this.#handshake;
    }

    /** Negotiate the workspace protocol for this connection. */
    async handshake(options: ClientOptions = {}): Promise<HandshakeResponse> {
        if (this.#handshake !== undefined) {
            return this.#handshake;
        }

        const response = await this.request(WorkspaceRequest.handshake({
            protocol: options.protocol ?? protocolRange,
            client: {
                ...clientDescriptor,
                ...options.client,
            },
            limits: options.limits ?? protocolLimits,
        }));

        if (response.kind === "handshake") {
            this.#handshake = response.handshake;

            return response.handshake;
        }

        if (response.kind === "error") {
            throw new Error(response.error.message);
        }

        throw new Error(`unexpected handshake response: ${response.kind}`);
    }

    /** Send one workspace request and wait for its response. */
    async request(
        payload: WorkspaceRequest,
        options: RequestOptions = {},
    ): Promise<WorkspaceResponse> {
        const previous = this.#requests;
        let release: () => void = () => {};
        const previousSettled = previous.catch(ignoreRequestFailure);
        this.#requests = previousSettled.then(() => new Promise<void>((resolve) => {
            release = resolve;
        }));

        await previousSettled;

        try {
            return await this.#request(payload, options);
        } finally {
            release();
        }
    }

    async #request(
        payload: WorkspaceRequest,
        options: RequestOptions,
    ): Promise<WorkspaceResponse> {
        if (this.#isClosed) {
            throw new Error("workspace connection is closed");
        }

        const id = this.#nextRequestId;
        this.#nextRequestId += 1;

        const message: ProtocolMessage = {
            kind: "request",
            request: {
                id: BigInt(id),
                options,
                payload,
            },
        };
        const response = new Promise<WorkspaceResponse>((resolve, reject) => {
            this.#pending.set(id, {
                resolve,
                reject,
                payloads: new PayloadReceiver(),
            });
        });

        try {
            await this.#transport.send(encodeFrame(message));
        } catch (error) {
            this.#pending.delete(id);
            throw error;
        }

        return response;
    }

    /** Close this connection. */
    close(): void {
        this.#isClosed = true;
        this.#transport.close();
        this.#rejectPending(new Error("workspace connection closed"));
    }

    async #receive(): Promise<void> {
        try {
            while (!this.#isClosed) {
                const frame = await this.#transport.receive();
                const message = decodeFrame(frame);
                this.#receiveMessage(message);
            }
        } catch (error) {
            const reason = error instanceof Error ? error : new Error(String(error));
            this.#rejectPending(reason);
        }
    }

    #receiveMessage(message: ProtocolMessage): void {
        if (message.kind === "response") {
            this.#receiveResponse(message.response);
        } else if (message.kind === "notification") {
            this.#receiveNotification(message.notification);
        } else {
            throw new Error("workspace connection received a request");
        }
    }

    #receiveResponse(response: ProtocolResponse): void {
        const id = checkedNumber(response.id, "response id");
        const pending = this.#pending.get(id);
        if (pending === undefined) {
            throw new Error(`workspace connection received unknown response id: ${id}`);
        }

        pending.response = response.payload;
        this.#resolvePending(id, pending);
    }

    #receiveNotification(notification: ProtocolNotification): void {
        if (notification.payload.kind === "payloadChunk") {
            this.#receivePayloadChunk(notification.payload.payload_chunk);
        } else {
            this.#onNotification?.(notification);
        }
    }

    #receivePayloadChunk(chunk: PayloadChunkNotification): void {
        const pending = this.#activePending();
        if (pending === undefined) {
            throw new Error("workspace connection received payload chunk without active request");
        }

        pending.entry.payloads.ingest(chunk);
        if (pending.entry.response !== undefined) {
            this.#resolvePending(pending.id, pending.entry);
        }
    }

    #activePending():
        | {
              readonly id: number;
              readonly entry: PendingResponse;
          }
        | undefined {
        const first = this.#pending.entries().next();
        if (first.done) {
            return undefined;
        }

        const [id, entry] = first.value;

        return { id, entry };
    }

    #resolvePending(id: number, pending: PendingResponse): void {
        if (pending.response === undefined) {
            return;
        }

        const response = pending.payloads.resolve(pending.response);
        if (response === undefined) {
            return;
        }

        this.#pending.delete(id);
        pending.resolve(response);
    }

    #rejectPending(error: Error): void {
        for (const pending of this.#pending.values()) {
            pending.reject(error);
        }

        this.#pending.clear();
    }
}

/** Ignore one earlier request failure while preserving request ordering. */
function ignoreRequestFailure(_error: unknown): void {}

/** Connect to one WebSocket workspace endpoint. */
export function connectEndpoint(
    url: string | URL,
    options: ConnectionOptions = {},
): Promise<Connection> {
    return Connection.connectWebSocket(url, options);
}
